### PR TITLE
Various fixes to handle wide characters correctly

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by lysnikolaou** [comment]
I messed up the base branch at the beginning. Sorry.

**Comment by lysnikolaou** [comment]
Marking as draft until I've added all the tests as well.

**Comment by pablogsal** [comment]
@lysnikolaou Can you add some tests to `setpos_from_xy` ?

File: Lib/_pyrepl/reader.py [hunk 1]
```diff
@@ -434,31 +437,44 @@ def push_input_trans(self, itrans) -> None:
...
+        j = 0
+        cur_x = self.screeninfo[i][0]
+        while cur_x < x:
+            if self.screeninfo[i][1][j] == 0:
+                continue
+            cur_x += self.screeninfo[i][1][j]
+            j += 1
+            pos += 1
+
+        self.pos = pos
```
Comment (pablogsal) [inline Lib/_pyrepl/reader.py:457]: What about something like (untested):

```
from itertools import chain, islice

def setpos_from_xy(self, x: int, y: int) -> None:
    """Set pos according to coordinates x, y"""
    pos = 0
    
    # Calculate the position based on the y coordinate
    row_lengths = (sum(cell for cell in row[1] if cell != 0) + 1 for row in islice(self.screeninfo, y))
    pos += sum(row_lengths)

    # Calculate the position based on the x coordinate
    cur_x, row_data = self.screeninfo[y]
    pos += sum(1 for width in row_data[:x] if width != 0)
    self.pos = pos
```

**Comment by pablogsal** [comment]
This is still a draft, are we waiting for the test before landing?

**Comment by lysnikolaou** [comment]
I'm currently working on adding tests for the console. Should we maybe land this and open a separate PR for that?

**Comment by pablogsal** [comment]
> I'm currently working on adding tests for the console. Should we maybe land this and open a separate PR for that?

Whatever works best for you. What do you prefer?

**Comment by lysnikolaou** [comment]
Yeah, let's merge this.

File: Lib/_pyrepl/reader.py [hunk 2]
```diff
@@ -434,31 +434,55 @@ def push_input_trans(self, itrans) -> None:
     def pop_input_trans(self) -> None:
         self.input_trans = self.input_trans_stack.pop()
 
-    def pos2xy(self, pos: int) -> tuple[int, int]:
+    def setpos_from_xy(self, x: int, y: int) -> None:
+        """Set pos according to coordincates x, y"""
```
Comment (pablogsal) [inline Lib/_pyrepl/reader.py:438]:
```suggestion
        """Set pos according to coordinates x, y"""
```

**Pull request merged**